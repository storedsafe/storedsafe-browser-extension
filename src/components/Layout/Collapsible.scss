@import '../../variables';

.collapsible {
  overflow: hidden;
  transition: max-height $transition-duration ease-out 0ms, visibility 0ms ease-out 0ms;

  &.collapsed {
    visibility: hidden;
    transition: max-height $transition-duration ease-out 0ms, visibility 0ms ease-out $transition-duration;
  }

  &.vertical.collapsed {
    max-height: 0;
  }

  &.horizontal.collapsed {
    max-width: 0;
  }
}

.collapsible.fixed {
  overflow: auto;
}
