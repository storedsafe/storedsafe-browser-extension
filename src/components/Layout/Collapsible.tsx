import * as React from 'react';
import './Collapsible.scss';

export interface CollapsibleProps {
  collapsed?: boolean;
  children: React.ReactNode;
  maxSize?: string;
  horizontal?: boolean;
}

export const Collapsible: React.FunctionComponent<CollapsibleProps> = ({
  collapsed,
  children,
  maxSize,
  horizontal,
}: CollapsibleProps) => {
  const collapsibleRef = React.useRef<HTMLDivElement>();
  const childRef = React.useRef<HTMLDivElement>();

  React.useEffect(() => {
    const child = childRef.current;
    const collapsible = collapsibleRef.current;

    const resize = (): void => {
      if (horizontal) {
        if (!collapsed) {
          if (maxSize) {
            collapsible.style.maxWidth = maxSize;
          } else {
            collapsible.style.maxWidth = `${child.clientWidth}px`;
          }
        } else {
          collapsible.style.maxWidth = '0';
        }
      } else {
        if (!collapsed) {
          if (maxSize) {
            collapsible.style.maxHeight = maxSize;
          } else {
            collapsible.style.maxHeight = `${child.clientHeight}px`;
          }
        } else {
          collapsible.style.maxHeight = '0';
        }
      }
    };

    resize();

    if (maxSize === undefined) {
      const resizeObserver = new ResizeObserver(() => resize());
      resizeObserver.observe(child);
      return (): void => resizeObserver.unobserve(child);
    }
  }, [collapsed, maxSize, horizontal]);

  const className = [
    'collapsible',
    horizontal ? 'horizontal' : 'vertical',
    maxSize === undefined ? '' : 'fixed',
    collapsed ? 'collapsed' : '',
  ].filter((name) => name !== '').join(' ');

  return (
    <div className={className} ref={collapsibleRef}>
      <div className="collapsible-children" ref={childRef}>
        {children}
      </div>
    </div>
  );
}

Collapsible.defaultProps = {
  collapsed: false,
  horizontal: false,
};
