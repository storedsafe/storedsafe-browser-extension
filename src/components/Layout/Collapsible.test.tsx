import * as React from 'react';
import { shallow } from 'enzyme';
import { Collapsible } from './Collapsible';

test('<Collapsible />', () => {
  const wrapper = shallow(<Collapsible>Children</Collapsible>);
  expect(wrapper).toMatchSnapshot();
});

test('<Collapsible collapsed=true />', () => {
  const wrapper = shallow(
    <Collapsible collapsed={true}>
      Children
    </Collapsible>
  );
  expect(wrapper).toMatchSnapshot();
});

test('<Collapsible collapsed=false />', () => {
  const wrapper = shallow(
    <Collapsible collapsed={false}>
      Children
    </Collapsible>
  );
  expect(wrapper).toMatchSnapshot();
});
