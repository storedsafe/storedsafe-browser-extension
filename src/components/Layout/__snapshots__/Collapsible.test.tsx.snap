// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible /> 1`] = `ShallowWrapper {}`;

exports[`<Collapsible collapsed=false /> 1`] = `ShallowWrapper {}`;

exports[`<Collapsible collapsed=true /> 1`] = `ShallowWrapper {}`;
